### PR TITLE
Resolve errors reconciling metrics-exporter

### PR DESCRIPTION
**Pull request by jarrpa**
Signed-off-by: Jose A. Rivera <jarrpa@redhat.com>

### CODE REVIEW

**Comment by obnoxxx** [comment]
@umangachapagain - PTAL.

Also this probably means this code is lacking tests?

File: pkg/controller/storagecluster/prometheus.go [hunk 1]
```diff
@@ -105,8 +105,9 @@ func (r *ReconcileStorageCluster) CreateOrUpdatePrometheusRules(rule *monitoring
 			if err != nil {
 				return fmt.Errorf("failed while updating PrometheusRule: %v", err)
 			}
+		} else {
+			return fmt.Errorf("failed while creating PrometheusRule: %v", err)
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/prometheus.go:109]: Why is this necessary?

File: pkg/controller/storagecluster/exporter.go [hunk 2]
```diff
@@ -105,6 +105,7 @@ func CreateOrUpdateService(r *ReconcileStorageCluster, instance *ocsv1.StorageCl
 		return nil, fmt.Errorf("failed to retrieve metrics exporter service %v. %v", namespacedName, err)
 	}
 	service.ResourceVersion = oldService.ResourceVersion
+	service.Spec.ClusterIP = oldService.Spec.ClusterIP
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/exporter.go:108]: I am not aware of Updates where we pass old ClusterIP to new Spec.
What issue is this causing?

**Review by umangachapagain — CHANGES_REQUESTED** [review]
I don't see any explanation of the error or how it's fixed. Makes it difficult to understand/review.
Please elaborate.

also ~metrics export~ -> metrics-exporter

**Comment by umangachapagain** [comment]
> @umangachapagain - PTAL.
> 
> Also this probably means this code is lacking tests?

It's tested along with the main reconcile block and was passing when it was merged.
But, we could always add more tests if we see a missing corner case.

File: pkg/controller/storagecluster/prometheus.go (hunk 1 above)
Comment (nbalacha) [reply pkg/controller/storagecluster/prometheus.go:109]: I think this looks good - otherwise it seems like this err will be returned every time the create fails with alreadyExists.

File: pkg/controller/storagecluster/exporter.go (hunk 2 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/exporter.go:108]: While the operator goes ready, we're flooding the logs with these error messages:

```

2020-10-14T17:32:25.967722321Z {"level":"error","ts":"2020-10-14T17:32:25.967Z","logger":"controller-runtime.controller","msg":"Reconciler error","controller":"storagecluster-controller","request":"openshift-storage/ocs-external-storagecluster","error":"failed to update service openshift-storage/ocs-metrics-exporter. Service \"ocs-metrics-exporter\" is invalid: spec.clusterIP: Invalid value: \"\": field is immutable","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/remote-source/app/vendor/github.com/go-logr/zapr/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/remote-source/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/remote-source/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/remote-source/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/remote-source/app/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/remote-source/app/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/remote-source/app/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```

Thus we need to maintain the same value for `ClusterIP`.

I would argue we shouldn't mark Ready if the metrics aren't set up properly, but that's a separate PR.

File: pkg/controller/storagecluster/prometheus.go (hunk 1 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/prometheus.go:109]: Correct. We want to ignore `err = alreadyExists`.

**Comment by jarrpa** [comment]
Updated the commit message.

File: pkg/controller/storagecluster/exporter.go (hunk 2 above)
Comment (umangachapagain) [reply pkg/controller/storagecluster/exporter.go:108]: I didn't want to block OCS usage just because we aren't able to set-up exporter (which is currently giving 1 or 2 metrics that we rarely use)

**Review by umangachapagain — APPROVED** [review]
